hedera-node/src/main/java/com/hedera/services/store/schedule/HederaScheduleStore.java

```diff
@@ -0,0 +1,228 @@
+import com.hedera.services.store.CreationResult;
+import com.hedera.services.store.HederaStore;
+import com.hederahashgraph.api.proto.java.AccountID;
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import com.hederahashgraph.api.proto.java.ScheduleID;
+import com.swirlds.fcmap.FCMap;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import static com.hedera.services.state.merkle.MerkleEntityId.fromScheduleId;
+import static com.hedera.services.store.CreationResult.failure;
+import static com.hedera.services.store.CreationResult.success;
+import static com.hedera.services.utils.EntityIdUtils.readableId;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_SCHEDULE_ACCOUNT_ID;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_SCHEDULE_ID;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_SCHEDULE_PAYER_ID;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.OK;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.SCHEDULE_WAS_DELETED;
+import static com.swirlds.common.CommonUtils.hex;
+
+/**
+ * Provides a managing store for Scheduled Entities.
+ *
+ * @author Daniel Ivanov
+ */
+public class HederaScheduleStore extends HederaStore implements ScheduleStore {
+	static final ScheduleID NO_PENDING_ID = ScheduleID.getDefaultInstance();
+
+	private final Supplier<FCMap<MerkleEntityId, MerkleSchedule>> schedules;
+	Map<String, MerkleEntityId> txHashToEntityId = new HashMap<>(); // HashMap<hash(txBytes), MerkleEntityId>
+
+	ScheduleID pendingId = NO_PENDING_ID;
+	String pendingTxHash = null;
+	MerkleSchedule pendingCreation;
+
+	public HederaScheduleStore(
+			EntityIdSource ids,
+			Supplier<FCMap<MerkleEntityId, MerkleSchedule>> schedules
+	) {
+		super(ids);
+		this.schedules = schedules;
+		buildTxBodyMap(this.schedules); // TODO: rebuild HashMap<hash(txBytes), {MerkleEntityId, List<AccountID>}>
+	}
+
+	@Override
+	public MerkleSchedule get(ScheduleID id) {
+		throwIfMissing(id);
+
+		return pendingId.equals(id) ? pendingCreation : schedules.get().get(fromScheduleId(id));
+	}
+
+	private void throwIfMissing(ScheduleID id) {
+		if (!exists(id)) {
+			throw new IllegalArgumentException(String.format("No such schedule '%s'!", readableId(id)));
+		}
+	}
+
+	@Override
+	public boolean exists(ScheduleID id) {
+		return pendingId.equals(id) || schedules.get().containsKey(fromScheduleId(id));
+	}
+
+	@Override
+	public void apply(ScheduleID id, Consumer<MerkleSchedule> change) {
+		throwIfMissing(id);
+		var key = fromScheduleId(id);
+		var schedule = schedules.get().getForModify(key);
+		Exception thrown = null;
+		try {
+			change.accept(schedule);
+		} catch (Exception e) {
+			thrown = e;
+		}
+		schedules.get().replace(key, schedule);
+		txHashToEntityId.remove(hex(schedule.transactionBody()));
```

Comment:
This line assumes that `change` is always deleting the scheduled tx although that is not the case (yes for now we are only deleting but potentially this could be expanded with something different than deletion). 
Maybe we could refactor it to:
```Java
if (schedule.isDeleted()) {
    txHashToEntityId.remove(hex(schedule.transactionBody()));
}
```

or even better might be to remove the default interface implementation and place it in the `HederScheduleStore`
```Java
ResponseCodeEnum delete(ScheduleID id) {
		var idRes = resolve(id);
		if (idRes == MISSING_SCHEDULE) {
			return INVALID_SCHEDULE_ID;
		}

		var schedule = get(id);
		if (schedule.adminKey().isEmpty()) {
			return SCHEDULE_IS_IMMUTABLE;
		}
		if (schedule.isDeleted()) {
			return SCHEDULE_WAS_DELETED;
		}

		apply(id, DELETION);
		txHashToEntityId.remove(hex(schedule.transactionBody()));
		return OK;
	}
```